platform/posix/transport/src/openssl_posix.c

```diff
@@ -729,55 +721,96 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
 {
     OpensslParams_t * pOpensslParams = NULL;
     int32_t bytesReceived = 0;
-    int32_t sslError = 0;
 
     if( ( pNetworkContext == NULL ) || ( pNetworkContext->pParams == NULL ) )
     {
         LogError( ( "Parameter check failed: pNetworkContext is NULL." ) );
     }
-    else if( pNetworkContext->pParams->pSsl != NULL )
+    else if( pNetworkContext->pParams->pSsl == NULL )
+    {
+        LogError( ( "Failed to receive data over network: "
+                    "SSL object in network context is NULL." ) );
+    }
+    else
     {
+        int32_t pollStatus = 1, readStatus = 1, sslError = 0;
+        uint8_t shouldRead = 0U;
+        struct pollfd pollFds;
         pOpensslParams = pNetworkContext->pParams;
-        /* SSL read of data. */
-        bytesReceived = ( int32_t ) SSL_read( pOpensslParams->pSsl,
-                                              pBuffer,
-                                              ( int32_t ) bytesToRecv );
+
+        /* Initialize the file descriptor. */
+        pollFds.events = POLLIN | POLLPRI | POLLHUP | POLLERR;
+        pollFds.revents = 0;
+        /* Set the file descriptor for poll. */
+        pollFds.fd = pOpensslParams->socketDescriptor;
+
+        if( ( bytesToRecv > 1 ) || ( SSL_pending( pOpensslParams->pSsl ) > 0 ) )
+        {
+            shouldRead = 1U;
+        }
+        else
+        {
+            /* Non-speculative read so check if there is data to read from the socket. */
+            pollStatus = poll( &pollFds, 1, 0 );
+        }
```

Comment:
Comments here to clearly mention that there is a non-blocking read for the case of reading more than one byte whereas  a timed wait when reading a single byte is the behavior represented by this if-else block

Reply:
`SSL_pending` will only return a non-zero value after a call to `SSL_read`. It just refers to data that was not yet processed on the last read. It is still necessary to have a mechanism to check if the socket has data to be read.

Reply:
Doesn't `SSL_pendign` data represent all the data buffered _after_ processing input records?

Reply:
I see a problem if a partial SSL record is received from the network, but `poll` returns a positive return code to represent availability of data on the TCP socket, but `SSL_read` returns zero bytes because the entire SSL record data has not been received. In that case, this function shouldn't return negative return code back to the calling MQTT/HTTP library

Reply:
We discussed that `poll` is necessary in this case because `SSL_pending` does not actually tell you if data is ready to be read from the socket, it will only return a value >0 if data remains to be read from the previous call to `SSL_read`.